captum/attr/_core/deep_lift.py

```diff
@@ -729,108 +726,122 @@ def _compute_mean_across_baselines(self, inp_bsz, base_bsz, attribution):
     )
     # normalizing
     n = np.prod(grad_input[0].shape)
-    grad_input[0] = grad_input_unnorm - grad_input_unnorm.sum() * 1 / n
-    return grad_input
 
+    # updating only the first half
+    new_grad_inp[0] = grad_input_unnorm - grad_input_unnorm.sum() * 1 / n
+    return new_grad_inp
 
-def maxpool1d(module, delta_in, delta_out, grad_input, grad_output, eps=1e-10):
+
+def maxpool1d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
     return maxpool(
         module,
         F.max_pool1d,
         F.max_unpool1d,
-        delta_in,
-        delta_out,
+        inputs,
+        outputs,
         grad_input,
         grad_output,
         eps=eps,
     )
 
 
-def maxpool2d(module, delta_in, delta_out, grad_input, grad_output, eps=1e-10):
+def maxpool2d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
     return maxpool(
         module,
         F.max_pool2d,
         F.max_unpool2d,
-        delta_in,
-        delta_out,
+        inputs,
+        outputs,
         grad_input,
         grad_output,
         eps=eps,
     )
 
 
-def maxpool3d(module, delta_in, delta_out, grad_input, grad_output, eps=1e-10):
+def maxpool3d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
     return maxpool(
         module,
         F.max_pool3d,
         F.max_unpool3d,
-        delta_in,
-        delta_out,
+        inputs,
+        outputs,
         grad_input,
         grad_output,
         eps=eps,
     )
 
 
 def maxpool(
-    module,
-    pool_func,
-    unpool_func,
-    delta_in,
-    delta_out,
-    grad_input,
-    grad_output,
-    eps=1e-10,
+    module, pool_func, unpool_func, inputs, outputs, grad_input, grad_output, eps=1e-10,
 ):
-    # The forward function of maxpool takes only tensors not
-    # a tuple hence accessing the first
-    # element in the tuple of inputs, grad_input and grad_output
-    _, indices = pool_func(
-        module.input[0],
-        module.kernel_size,
-        module.stride,
-        module.padding,
-        module.dilation,
-        module.ceil_mode,
-        True,
-    )
-    unpool_grad_out_delta = unpool_func(
-        grad_output[0] * delta_out[0],
-        indices,
-        module.kernel_size,
-        module.stride,
-        module.padding,
-        list(module.input[0].shape),
-    )
+    with torch.no_grad():
+        input, input_ref = inputs.chunk(2)
+        output, output_ref = outputs.chunk(2)
+
+        delta_in = input - input_ref
+        delta_in = torch.cat(2 * [delta_in])
+        # Extracts cross maximum between the outputs of maxpool for the
+        # actual inputs and its corresponding references. In case the delta outputs
+        # for the references are larger the method relies on the references and
+        # corresponding gradients to compute the multiplies and contributions.
+        delta_out_xmax = torch.max(output, output_ref)
+        delta_out = torch.cat([delta_out_xmax - output_ref, output - delta_out_xmax])
```

Comment:
Good point! 2nd option is good however it doesn't work with DataParallel because the input and the baseline will be concatenated in a different order.